File: pkg/@eserstack/workflows/registry.ts
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

/**
 * Tool registry — a container for workflow tools, looked up by name.
 *
 * @example
 * ```typescript
 * import * as workflows from "@eserstack/workflows";
 *
 * const registry = workflows.createRegistry();
 * registry.register({ name: "fix-eof", description: "...", run: async () => ({...}) });
 * registry.get("fix-eof");
 * ```
 *
 * @module
 */

import type { WorkflowTool } from "./types.ts";

/** A tool registry instance. */
export type Registry = {
  /** Register a single tool. */
  readonly register: (tool: WorkflowTool) => void;
  /** Register multiple tools at once. */
  readonly registerAll: (tools: readonly WorkflowTool[]) => void;
  /** Get a tool by name. */
  readonly get: (name: string) => WorkflowTool | undefined;
  /** Get all registered tools. */
  readonly getAll: () => readonly WorkflowTool[];
  /** Check if a tool is registered. */
  readonly has: (name: string) => boolean;
  /** Get all registered tool names. */
  readonly names: () => readonly string[];
};

/**
 * Create a new tool registry.
 *
 * @returns An empty Registry instance
 */
export const createRegistry = (): Registry => {
  const tools = new Map<string, WorkflowTool>();

  return {
    register: (tool) => {
      tools.set(tool.name, tool);
    },
    registerAll: (toolList) => {
      for (const tool of toolList) {
        tools.set(tool.name, tool);
      }
    },
    get: (name) => tools.get(name),
    getAll: () => [...tools.values()],
    has: (name) => tools.has(name),
    names: () => [...tools.keys()],
  };
};

File: pkg/ajan/workflowfx/registry.go
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

package workflowfx

import (
	"fmt"
	"sort"
	"sync"
)

// Registry stores WorkflowTool implementations by name.
type Registry struct {
	tools map[string]WorkflowTool
	mu    sync.RWMutex
}

// NewRegistry creates an empty tool registry.
func NewRegistry() *Registry {
	return &Registry{
		tools: make(map[string]WorkflowTool),
		mu:    sync.RWMutex{},
	}
}

// Register adds a tool. Overwrites any existing tool with the same name.
func (r *Registry) Register(tool WorkflowTool) {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.tools[tool.Name()] = tool
}

// Unregister removes a tool by name.
func (r *Registry) Unregister(name string) {
	r.mu.Lock()
	defer r.mu.Unlock()

	delete(r.tools, name)
}

// Get looks up a tool by name.
func (r *Registry) Get(name string) (WorkflowTool, bool) { //nolint:ireturn
	r.mu.RLock()
	defer r.mu.RUnlock()

	tool, ok := r.tools[name]

	return tool, ok
}

// Names returns a sorted list of registered tool names.
func (r *Registry) Names() []string {
	r.mu.RLock()
	defer r.mu.RUnlock()

	names := make([]string, 0, len(r.tools))

	for name := range r.tools {
		names = append(names, name)
	}

	sort.Strings(names)

	return names
}

// MustGet returns a tool or panics if not found. Useful in test setup.
func (r *Registry) MustGet(name string) WorkflowTool { //nolint:ireturn
	tool, ok := r.Get(name)
	if !ok {
		panic(fmt.Sprintf("workflowfx: tool '%s' not registered", name))
	}

	return tool
}
